### PR TITLE
[java] Update rule UseVarargs

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

Update rule UseVarargs. I think this is the last one of bestpractices, excluding UnusedImports, which I'd like to port much later.

## Related issues

- Item of #2701 

## Ready?

- [ ] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [ ] Added (in-code) documentation (if needed)

### CODE REVIEW

File: pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/Java16TreeDumpTest.java [hunk 1]
```diff
@@ -11,22 +11,18 @@
...
-public class Java16TreeDumpTest extends BaseTreeDumpTest {
+public class Java16TreeDumpTest extends BaseJavaTreeDumpTest {
     private final JavaParsingHelper java16 =
             JavaParsingHelper.WITH_PROCESSING.withDefaultVersion("16")
                     .withResourceContext(Java15TreeDumpTest.class, "jdkversiontests/java16/");
     private final JavaParsingHelper java16p = java16.withDefaultVersion("16-preview");
     private final JavaParsingHelper java15 = java16.withDefaultVersion("15");
 
-    public Java16TreeDumpTest() {
-        super(new RelevantAttributePrinter(), ".java");
```
Comment (oowekyala) [inline pmd-java/src/test/java/net/sourceforge/pmd/lang/java/ast/Java16TreeDumpTest.java:28]: Note that this is why the reference files have changed that much. The printer of the new superclass filters out deprecated attributes, among other things.

File: pmd-java/src/main/resources/category/java/bestpractices.xml [hunk 2]
```diff
@@ -1751,29 +1751,12 @@ having to deal with the creation of an array.
...
-             and child::ResultType[@Void=true()]
-             and @Name = 'main'
-             and @Arity = 1
-            ]
-         (: Type of the formal parameter here. :)
-         and pmd-java:typeIs('java.lang.String[]')
-    )]
+//FormalParameters[not(parent::MethodDeclaration[@Overridden=true() or @MainMethod=true()])]
+  /FormalParameter[position()=last()]
+   [ArrayType[not(PrimitiveType or ClassOrInterfaceType[pmd-java:typeIs('java.lang.Byte')])] and @Varargs=false()]
```
Comment (oowekyala) [inline pmd-java/src/main/resources/category/java/bestpractices.xml:1759]: Now it allows all primitive arrays instead of just `byte[]`. This is assymetric, I'll probably have to change that later...

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 0 violations,<br />
introduces 330 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/2744c73b14b74290bd298ef496e9095d/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 2343 violations,<br />
introduces 5932 new violations, 1 new errors and 0 new configuration errors,<br />
removes 15474 violations, 10 errors and 2 configuration errors.<br />
<a href="https://chunk.io/pmd/2744c73b14b74290bd298ef496e9095d/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/resources/category/java/bestpractices.xml (hunk 2 above)
Comment (oowekyala) [reply pmd-java/src/main/resources/category/java/bestpractices.xml:1759]: How about we just remove this exception? I don't see why `byte[]` should be special cased over eg `int[]`. In some kinds of code byte arrays are common, but that's just the same for any array...

Comment (adangel) [reply pmd-java/src/main/resources/category/java/bestpractices.xml:1759]: I wouldn't change it too much. If we change it, it deserves a proper issue/PR by its own.
I can't tell you, why it was that way it is, but that's it is now. And this PR is about making the rule run with PMD 7 and not improving the rule what so ever....

I see now a couple of false negatives:
* https://github.com/checkstyle/checkstyle/tree/checkstyle-8.10/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule4832nocstylearray/InputArrayTypeStyle.java#L15
    * This I don't know why we don't flag it anymore. It is not the main method. Maybe, because the other array notation is used (`[]` at variableid rather than at the type...)
    * Same case: https://github.com/checkstyle/checkstyle/tree/checkstyle-8.10/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfter.java#L275
* https://github.com/checkstyle/checkstyle/tree/checkstyle-8.10/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/needbraces/InputNeedBracesSingleLineStatements.java#L99
    * I think, we should keep the old behavior and flag `int[]` as before
    * Same case, but with `char`: https://github.com/checkstyle/checkstyle/tree/checkstyle-8.10/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterArrayDeclarations3.java#L8
* https://github.com/checkstyle/checkstyle/tree/checkstyle-8.10/src/test/resources/com/puppycrawl/tools/checkstyle/checks/uncommentedmain/InputUncommentedMain4.java#L10
    * That's also not a main method and should be flagged


This removed violation is debatable:
* https://github.com/checkstyle/checkstyle/tree/checkstyle-8.10/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/nowhitespaceafter/InputNoWhitespaceAfterArrayDeclarations2.java#L35
    * That's a very weird notation of arrays. With PMD 6, we also reported this, since you can write a 2-dim array like that: `int[] ... a`.
    * Not sure, if it really makes sense to suggest this, but PMD 6 did it that way.
    * _Update:_ This is probably not reported anymore, because all primitive types are excluded...

The new violations look all correct.
* https://github.com/checkstyle/checkstyle/tree/checkstyle-8.10/src/test/resources/com/puppycrawl/tools/checkstyle/grammars/java8/InputAnnotationsOnArray.java#L23
    * This one is a case for a 2-dim array. So the argument above is non-sense. It's about primitive types and not about dims.

In summary: I'd just exclude `byte[]` and keep flagging other primitive arrays as it was before.

Comment (oowekyala) [reply pmd-java/src/main/resources/category/java/bestpractices.xml]: >  And this PR is about making the rule run with PMD 7 and not improving the rule what so ever....

I'm not sure how to feel about this statement... I think if we don't seize opportunities to improve simple rules when the overhead would be minimal, we won't get through our stack of todo rules without spawning more enhancement tickets that will get buried than we actually fix issues.

OTOH maybe this specific issue could also be done on master. It's just that it's so much overhead to write an enhancement ticket, open a pr, deal with the merge into pmd 7, etc... I'd like to spend our resources on actual features and not just bureaucracy surrounding fixing legacy, undocumented behavior.

Comment (adangel) [reply pmd-java/src/main/resources/category/java/bestpractices.xml]: >  It's just that it's so much overhead to write an enhancement ticket, open a pr, deal with the merge into pmd 7,

IMHO this is required for an open source project. If it's a private project, nobody cares. Writing an enhancement ticket is documentation. While converting the rule you basically found a new bug, but you started out to convert the rule and not to fix this bug in the first place.
Since you already fixed this new bug in this PR, all that is left, is to describe, what bug you fixed in a issue and mention it in the release notes for PMD 7.

Comment (adangel) [reply pmd-java/src/main/resources/category/java/bestpractices.xml]: Ok, so the rule is back to it's state like it is currently on master.

The question now is: Why does this rule handle `byte[]` different than e.g. `int[]`? I would ignore this problem for now completely until some user actually reports this.

Maybe we should just document this exception in the rule description. I can only guess, that working with bytes en bloc as an array is more common than dealing with separate individual byte values. Probably this is from times, when java.nio.ByteBuffer didn't exist yet.

Originally, when the rule was introduced in 5.0.0, it didn't have an exception for byte arrays. The exception was added with 5.2.3 (https://pmd.sourceforge.io/pmd-5.2.3/overview/changelog.html) via #46 and #47.

**Review by adangel — APPROVED** [review]
